# Exploring Medical Payments

In this activity, you’ll explore medical payments for diabetes by state. You’ll do so by using interactive visualizations, numerical and visual aggregation, and analysis.


Instructions:

Note that data from `hospital_claims.csv` has already been loaded into this notebook, and three DataFrames have been created: 

* `hospital_data` contains all of the information from the CSV file.

* `procedure_638_payments` contains all columns of data with a "DRG Definition" of "638 - DIABETES W CC"

* `average_total_payments_by_state` contains only the "Provider State" and "Average Total Payments" columns from the `procedure_638_payments` DataFrame.

This activity is broken down into three main sections:

1. Explore the average total payments by state

2. Explore the average total payments in California

3. Explore the outliers in the data

In the first two sections, you’ll both numerically and visually aggregate the data from the DataFrame that the starter code supplies.

#### Explore the Average Total Payments by State

In this section, you’ll explore the average total payments by state by numerically and visually aggregating the data. 

##### Numerically Aggregate the Data

First, apply grouping and numerical aggregation by completing the following steps:

1. Create a DataFrame called `total_payments_by_state` by grouping the `average_total_payments_by_state` DataFrame by “Provider State”. Then use the `sum` function to aggregate the results.

2. Sort the data by price by using `.sort_values (“Average Total Payments”)` on the DataFrame from Step 1.

3. Display the first five states in the sorted DataFrame. These states have the lowest total payments.

4. Display the state with the highest total payment.

##### Visually Aggregate the Data

Next, use hvPlot to create interactive visualizations that allow you to explore the data. To do so, complete the following steps:

1. Import the hvPlot library for Pandas.

2. Plot the `total_payments_by_state` DataFrame by using the following code:
    
    ```python
    total_payments_by_state_sorted.hvplot.bar(x="Provider State", y="Average Total Payments")
    ```

3. Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

    * Based on your visualization, which state has the highest total payment?
    * Which state has the lowest total payment?

#### Explore the Average Total Payments in California

In the previous section, you found that California has the highest total payments for diabetes care. In this section, you’ll further explore the California prices by using visual and numerical aggregation.

##### Visually Aggregate the Data

First, visually aggregate the data by completing the following steps:

1. Using the `average_total_payments_by_state` DataFrame, create a bar plot, and group the data by provider state via the hvPlot `groupby` parameter. Use the hvPlot widget dropdown menu to select the prices for California.

2. Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

    * What patterns do you notice in the data?
    * Do any data points influence the total payments amount?

##### Numerically Aggregate the Data

Next, numerically aggregate the California data by completing the following steps: 

1. Create a DataFrame by using the `loc` function and by using the `average_total _payments_by_state` DataFrame filtered only for California (CA). Sort the values by average total payments, and then review both the first five and the last five rows of data.

2. Plot the sorted DataFrame as an hvPlot bar chart.

3. Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:
    * How many payments appear to be outliers?
    * What’s the approximate value of those payments?

#### Explore the Outliers in the Data

In this section, you’ll use the DataFrame that you generated in your analysis of the California average total payments to explore the outlier payments. Complete the following steps:

1. Use the `describe` function to find the mean value of the average total payments.

2. Use the interactive bar plot (for the sorted values of the California average total payments) to estimate a payment value that you can use to filter out the highest three data spikes.

3. Use `loc` or `iloc` to filter out the three outlier payments from the California average total payments. Then recalculate the summary statistics by using the `describe` function.

4. Review the two sets of summary statistics that you generated in this section and answer the following question. 

    * How much do the outliers change the mean value of the California average total payments for diabetes?



References:

[Pandas groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html)

In [21]:
# Import the required libraries and dependencie

import plotly.express as px
import pandas as pd
import os
from pathlib import Path
from dotenv import load_dotenv

## Prepare the Data (Included)

Three DataFrames have been created: 

* `hospital_data` contains all of the information from the CSV file.

* `procedure_638_payments` contains all columns of data with a "DRG Definition" of "638 - DIABETES W CC"

* `average_total_payments_by_state` contains only the "Provider State" and "Average Total Payments" columns from the `procedure_638_payments` DataFrame.

In [22]:
# Using the read_csv function and Path module, read in the 
# hospital_claims.csv file from the Resources folder
hospital_data = pd.read_csv(Path("../Resources/hospital_claims.csv"))

# Review the DataFrame
hospital_data.head()

DRG Definition  Provider Id  \
0  039 - EXTRACRANIAL PROCEDURES W/O CC/MCC        10001   
1  039 - EXTRACRANIAL PROCEDURES W/O CC/MCC        10005   
2  039 - EXTRACRANIAL PROCEDURES W/O CC/MCC        10006   
3  039 - EXTRACRANIAL PROCEDURES W/O CC/MCC        10011   
4  039 - EXTRACRANIAL PROCEDURES W/O CC/MCC        10016   

                      Provider Name     Provider Street Address Provider City  \
0  SOUTHEAST ALABAMA MEDICAL CENTER      1108 ROSS CLARK CIRCLE        DOTHAN   
1     MARSHALL MEDICAL CENTER SOUTH  2505 U S HIGHWAY 431 NORTH          BOAZ   
2    ELIZA COFFEE MEMORIAL HOSPITAL          205 MARENGO STREET      FLORENCE   
3                 ST VINCENT'S EAST  50 MEDICAL PARK EAST DRIVE    BIRMINGHAM   
4     SHELBY BAPTIST MEDICAL CENTER     1000 FIRST STREET NORTH     ALABASTER   

  Provider State  Provider Zip Code Hospital Referral Region Description  \
0             AL              36301                          AL - Dothan   
1             AL              35957                      AL - Birmingham   
2             AL              35631                      AL - Birmingham   
3             AL              35235                      AL - Birmingham   
4             AL              35007                      AL - Birmingham   

   Total Discharges  Average Covered Charges  Average Total Payments  \
0                91                 32963.07                 5777.24   
1                14                 15131.85                 5787.57   
2                24                 37560.37                 5434.95   
3                25                 13998.28                 5417.56   
4                18                 31633.27                 5658.33   

   Average Medicare Payments  
0                    4763.73  
1                    4976.71  
2                    4453.79  
3                    4129.16  
4                    4851.44

In [23]:
# Slice the data to only look at medical costs associated with diabetes
# the DRG Definition code is "638 - DIABETES W CC"
procedure_638_payments = hospital_data.loc[hospital_data["DRG Definition"] == "638 - DIABETES W CC"]

# Review the first and last five rows of the DataFrame
display(procedure_638_payments.head())
display(procedure_638_payments.tail())

DRG Definition  Provider Id                     Provider Name  \
125166  638 - DIABETES W CC       330004                 KINGSTON HOSPITAL   
127575  638 - DIABETES W CC        10001  SOUTHEAST ALABAMA MEDICAL CENTER   
127576  638 - DIABETES W CC        10005     MARSHALL MEDICAL CENTER SOUTH   
127577  638 - DIABETES W CC        10006    ELIZA COFFEE MEMORIAL HOSPITAL   
127578  638 - DIABETES W CC        10011                 ST VINCENT'S EAST   

           Provider Street Address Provider City Provider State  \
125166                396 BROADWAY      KINGSTON             NY   
127575      1108 ROSS CLARK CIRCLE        DOTHAN             AL   
127576  2505 U S HIGHWAY 431 NORTH          BOAZ             AL   
127577          205 MARENGO STREET      FLORENCE             AL   
127578  50 MEDICAL PARK EAST DRIVE    BIRMINGHAM             AL   

        Provider Zip Code Hospital Referral Region Description  \
125166              12401                          NY - Albany   
127575              36301                          AL - Dothan   
127576              35957                      AL - Birmingham   
127577              35631                      AL - Birmingham   
127578              35235                      AL - Birmingham   

        Total Discharges  Average Covered Charges  Average Total Payments  \
125166                21                 20006.57                 6048.85   
127575                32                 21175.81                 4678.43   
127576                12                  9719.16                 4863.75   
127577                35                 17021.54                 4434.57   
127578                14                 15875.50                 5176.07   

        Average Medicare Payments  
125166                    5177.85  
127575                    4047.68  
127576                    4203.41  
127577                    3537.20  
127578                    3394.64

DRG Definition  Provider Id  \
129389  638 - DIABETES W CC       670031   
129390  638 - DIABETES W CC       670041   
129391  638 - DIABETES W CC       670047   
129392  638 - DIABETES W CC       670055   
129393  638 - DIABETES W CC       670056   

                                Provider Name  \
129389      ST LUKE'S PATIENTS MEDICAL CENTER   
129390        SETON MEDICAL CENTER WILLIAMSON   
129391  SIERRA PROVIDENCE EAST MEDICAL CENTER   
129392           METHODIST STONE OAK HOSPITAL   
129393              SETON MEDICAL CENTER HAYS   

                    Provider Street Address Provider City Provider State  \
129389  4600 EAST SAM HOUSTON PARKWAY SOUTH      PASADENA             TX   
129390                    201 SETON PARKWAY    ROUND ROCK             TX   
129391                 3280 JOE BATTLE BLVD       EL PASO             TX   
129392                 1139 E SONTERRA BLVD   SAN ANTONIO             TX   
129393                       6001 KYLE PKWY          KYLE             TX   

        Provider Zip Code Hospital Referral Region Description  \
129389              77505                         TX - Houston   
129390              78664                          TX - Austin   
129391              79938                         TX - El Paso   
129392              78258                     TX - San Antonio   
129393              78640                          TX - Austin   

        Total Discharges  Average Covered Charges  Average Total Payments  \
129389                14                 17782.71                 4610.00   
129390                11                 34838.54                 4596.27   
129391                15                 32414.26                 4943.06   
129392                13                 35024.23                 6881.30   
129393                12                 37231.33                 8237.41   

        Average Medicare Payments  
129389                    3965.42  
129390                    3884.63  
129391                    4190.53  
129392                    3900.15  
129393                    5143.66

In [24]:
# Create a DataFrame that includes only the Provider State and Average Total Payments column
# from the procedure 638 information
average_total_payments_by_state = procedure_638_payments[["Provider State", "Average Total Payments"]]

# Review the first and last five rows of the DataFrame
display(average_total_payments_by_state.head())
display(average_total_payments_by_state.tail())

Provider State  Average Total Payments
125166             NY                 6048.85
127575             AL                 4678.43
127576             AL                 4863.75
127577             AL                 4434.57
127578             AL                 5176.07

Provider State  Average Total Payments
129389             TX                 4610.00
129390             TX                 4596.27
129391             TX                 4943.06
129392             TX                 6881.30
129393             TX                 8237.41

## Explore Average Total Payments by State

In this section, you’ll explore the average total payments by state by numerically and visually aggregating the data. 

### Numerically Aggregate the Data

First, apply grouping and numerical aggregation by completing the following steps:

1. Create a DataFrame called `total_payments_by_state` by grouping the `average_total_payments_by_state` DataFrame by “Provider State”. Then use the `sum` function to aggregate the results.

2. Sort the data by price by using `.sort_values (“Average Total Payments”)` on the DataFrame from Step 1.

3. Display the first five states in the sorted DataFrame. These states have the lowest total payments.

4. Display the state with the highest total payment.



#### Step 1: Create a DataFrame called `total_payments_by_state` by grouping the `average_total_payments_by_state` DataFrame by “Provider State”. Then use the `sum` function to aggregate the results.

In [25]:
# Group the data by the "Provider State" and then sum the aggregate
total_payments_by_state = average_total_payments_by_state.groupby('Provider State').sum()

# View the result
# YOUR CODE HERE
display(total_payments_by_state.head())

Average Total Payments
Provider State                        
AK                            15525.04
AL                           194535.26
AR                           111746.03
AZ                           158308.98
CA                          1201241.89

#### Step 2: Sort the data by price by using `.sort_values (“Average Total Payments”)` on the DataFrame from Step 1.

In [26]:
# Sort the states by their price
# and display the top and bottom five states by total payments
total_payments_by_state_sorted = total_payments_by_state.sort_values("Average Total Payments")
display(total_payments_by_state_sorted.head())
display(total_payments_by_state_sorted.tail())

Average Total Payments
Provider State                        
WY                            13610.80
HI                            13904.54
AK                            15525.04
SD                            17308.57
VT                            20287.08

Average Total Payments
Provider State                        
IL                           548774.53
FL                           658981.35
TX                           719401.64
NY                           871073.79
CA                          1201241.89

#### Step 3: Display the first five states in the sorted DataFrame. These states have the lowest total payments.

In [27]:
# Display the first 5 states in the sorted DataFrame.
# These should have the lowest total payments
# YOUR CODE HERE
display(total_payments_by_state_sorted.head())


Average Total Payments
Provider State                        
WY                            13610.80
HI                            13904.54
AK                            15525.04
SD                            17308.57
VT                            20287.08

#### Step 4: Display the state with the highest total payment.

In [28]:
# Determine which state has the highest average total payment
state_with_highest_total = total_payments_by_state_sorted.tail(1)


# View the state
# YOUR CODE HERE
state_with_highest_total

Average Total Payments
Provider State                        
CA                          1201241.89

---

### Visually Aggregate the Data

Next, use hvPlot to create interactive visualizations that allow you to explore the data. To do so, complete the following steps:

1. Import the hvPlot library for Pandas.

2. Plot the `total_payments_by_state` DataFrame by using the following code:
    
    ```python
    total_payments_by_state_sorted.hvplot.bar(x="Provider State", y="Average Total Payments")
    ```

3. Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

    * Based on your visualization, which state has the highest total payment?

    * Which state has the lowest total payment?





#### Step 1: Import the hvPlot library for Pandas.

In [29]:
# Import the hvPlot library
# YOUR CODE HERE

import hvplot.pandas

#### Step 2: Plot the `total_payments_by_state` DataFrame by using the following code:
    
    ```python
    total_payments_by_state_sorted.hvplot.bar(x="Provider State", y="Average Total Payments")
    ```

In [30]:
# Create an interactive bar plot with hvplot where the x-axis is the "Provider State" 
# and the y-axis is the "Average Total Payments"
# YOUR CODE HERE

total_payments_by_state_sorted.hvplot.bar(
    x="Provider State",
    y="Average Total Payments",
    rot=90
)

:Bars   [Provider State]   (Average Total Payments)

#### Step 3: Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

**Question** Based on your visualization, which state has the highest total payment?

**Answer**  # YOUR ANSWER HERE


**Question** Which state has the lowest total payment?

**Answer** # YOUR ANSWER HERE

---

## Explore the Average Total Payments in California

In the previous section, you found that California has the highest total payments for diabetes care. In this section, you’ll further explore the California prices by using visual and numerical aggregation.


### Visually Aggregate the Data

First, visually aggregate the data by completing the following steps:

1. Using the `average_total_payments_by_state` DataFrame, create a bar plot, and group the data by provider state via the hvPlot `groupby` parameter. Use the hvPlot widget dropdown menu to select the prices for California.

2. Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

    * What patterns do you notice in the data?

    * Do any data points influence the total payments amount?


#### Step 1: Using the `average_total_payments_by_state` DataFrame, create a bar plot, and group the data by provider state via the hvPlot `groupby` parameter. Use the hvPlot widget dropdown menu to select the prices for California.

In [31]:
# Create a hvplot.bar chart and groupby the "Provider State"
# then select California
# YOUR CODE HERE


average_total_payments_by_state.hvplot.bar(
    groupby="Provider State",
    rot=90,
    size=(35, 7)
)

:DynamicMap   [Provider State]
   :Bars   [index]   (Average Total Payments)

#### Step 2: Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

**Question** Referring to the visualization of California's data, what patterns do you notice in the data?

**Answer** # YOUR ANSWER HERE



**Question** Do any data points influence the total payments amount?

**Answer**  # YOUR ANSWER HERE

### Numerically Aggregate the Data

Next, numerically aggregate the California data by completing the following steps: 

1. Create a DataFrame by using the `loc` function and by using the `average_total _payments_by_state` DataFrame filtered only for California (CA). Sort the values by average total payments, and then review both the first five and the last five rows of data.

2. Plot the sorted DataFrame as an hvPlot bar chart.

3. Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:
    * How many payments appear to be outliers?
    * What’s the approximate value of those payments?


#### Step 1: Create a DataFrame by using the `loc` function and by using the `average_total _payments_by_state` DataFrame filtered only for California (CA). Sort the values by average total payments, and then review both the first five and the last five rows of data.

In [32]:
# Using the `loc` function, create a DataFrame of the average total payment by state
# filtered for only California (CA)
ca_average_total_payments = average_total_payments_by_state.loc[average_total_payments_by_state['Provider State']=="CA"]

# Sort the data
ca_average_total_payments = ca_average_total_payments.sort_values("Average Total Payments")

# View the first five rows of the DataFrame
display(ca_average_total_payments.head())
display(ca_average_total_payments.tail())
    

Provider State  Average Total Payments
127752             CA                 4912.60
127733             CA                 4968.57
127721             CA                 5060.46
127800             CA                 5237.00
127760             CA                 5239.00

Provider State  Average Total Payments
127676             CA                12598.66
127781             CA                12793.37
127678             CA                15566.14
127746             CA                15731.08
127735             CA                19511.95

#### Step 2: Plot the sorted DataFrame as an hvPlot bar chart.

In [33]:
# Use hvPlot to visualize the sorted data for California
# YOUR CODE HERE
ca_average_total_payments.hvplot.bar(rot=90, size=(35,7))

:Bars   [index]   (Average Total Payments)

#### Step 3: Use your visualization and apply the pan, zoom, and hover widgets to answer the following questions:

**Question** How many payments appear to be outliers? What’s the approximate value of those payments?

**Answer** Zooming in on the data, it appears that there are 3 payments that appear to be outliers. The approximate value of these outliers are: a payment of approximately 15,570 at index position 127678; a payment of approximately 15,730 at index position 127746; and a payment of approximately 19,510 at index position 127735. 

### Explore the Outliers in the Data

In this section, you’ll use the DataFrame that you generated in your analysis of the California average total payments to explore the outlier payments. Complete the following steps:

1. Use the `describe` function to find the mean value of the average total payments.

2. Use the interactive bar plot (for the sorted values of the California average total payments) to estimate a payment value that you can use to filter out the highest three data spikes.

3. Use `loc` or `iloc` to filter out the three outlier payments from the California average total payments. Then recalculate the summary statistics by using the `describe` function.

4. Review the two sets of summary statistics that you generated in this section and answer the following question. 

* How much do the outliers change the mean value of the California average total payments for diabetes?



#### Step 1: Use the `describe` function to find the mean value of the average total payments.

In [34]:
# Use the describe function to calculate the mean value of California's average total payments.
# YOUR CODE HERE
ca_average_total_payments.describe()

Average Total Payments
count              160.000000
mean              7507.761812
std               1962.283033
min               4912.600000
25%               6269.972500
50%               7016.330000
75%               8358.897500
max              19511.950000

#### Step 2: Use the interactive bar plot (for the sorted values of the California average total payments) to estimate a payment value that you can use to filter out the highest three data spikes.

In [35]:
# Using the ca_average_total_payments DataFrame, create a conditional statement 
# that can be used to filter out the three largest payments
# YOUR CODE HERE

ca_average_total_payments["Average Total Payments"] < 150000

127752    True
127733    True
127721    True
127800    True
127760    True
          ... 
127676    True
127781    True
127678    True
127746    True
127735    True
Name: Average Total Payments, Length: 160, dtype: bool

#### Step 3: Use `loc` to filter out the three outlier payments from the California average total payments. Then recalculate the summary statistics by using the `describe` function.

In [36]:
# Create a DataFrame that filters out the 3 largest payments from the California data 
filtered_california_payments = ca_average_total_payments.loc[
    ca_average_total_payments["Average Total Payments"] < 500000
]

# View the filtered DataFrame in a plot
# YOUR CODE HERE
filtered_california_payments.hvplot.bar(rot=90, size=(35, 7))

:Bars   [index]   (Average Total Payments)

In [37]:
# Use the describe function to calculate summary statistics for the filtered data.
# YOUR CODE HERE
['Provider State', 'Average Total Payments'] but received: PopulationCount

In [38]:
filtered_california_payments.describe()

Average Total Payments
count              160.000000
mean              7507.761812
std               1962.283033
min               4912.600000
25%               6269.972500
50%               7016.330000
75%               8358.897500
max              19511.950000

In [45]:
# Plot data using Plotly and the Mapbox scatter plot
px.scatter_mapbox(
    filtered_california_payments,
    lat='Provider State',
    lon='Average Total Payments',
    size='Provider State',
    color='Average Total Payments',
    zoom=4,
    height=750
)

TypeError: unsupported operand type(s) for /: 'str' and 'int'

#### Step 4: Review the two sets of summary statistics that you generated in this section and answer the following question.

**Question**  How much do the outliers change the mean value of the California average total payments for diabetes?

**Answer** # YOUR ANSWER HERE